# Hola! 😀

Soy **Jaime Paz** y mis amigos me suelen llamar James! – *"Tranquilo, no juego en el Real Madrid"* ⚽😅. Sí, como el futbolista James Rodríguez, pero en versión científico de datos… y mejor no me digas como a él. 😉

Como tu revisor en TripleTen, estoy aquí para ayudarte a pulir tu código y tu forma de trabajar con datos. Si algo necesita un ajuste, no te preocupes: la idea es que cada comentario te acerque más a cómo se trabaja en un entorno profesional y que tu proyecto brille con todo su potencial.

Cada vez que encuentre un detalle importante en tu notebook, te lo señalaré para que puedas corregirlo y seguir creciendo. Si en algún punto no logras resolver algo, también estoy para guiarte en próximos intentos de revisión. 🔁

Es muy importante que, cuando veas mis comentarios en el notebook, **no los muevas, no los modifiques y no los borres**. Así mantenemos un historial claro de lo que ya revisamos y de tus avances. ✅

---

### Formato de Comentarios

Revisaré cuidadosamente tu notebook para asegurar que cumpla con los requisitos y te daré comentarios usando el siguiente formato:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> ✅ - ¡Excelente trabajo! Esta parte está bien implementada y contribuye de forma positiva al análisis o al proyecto. Sigue aplicando estas buenas prácticas en las siguientes secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Esta parte del código funciona, pero se puede mejorar u optimizar. Tal vez sea más claro, más eficiente o más fácil de mantener. Te señalaré ideas para que puedas reforzar esta sección.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error que es necesario corregir para aprobar esta parte. Revisa el comentario con calma, ajústalo y vuelve a intentarlo; es clave para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto. 💪
- **Áreas de mejora:** Sugerencias claras sobre lo que puedes reforzar. 💡
- **Temas adicionales para investigar:** Ideas opcionales que puedes explorar por tu cuenta para seguir creciendo.

Estos temas adicionales no son obligatorios ahora, pero pueden ayudarte a profundizar en el futuro. 📚

---

Si tienes dudas o quieres responder a un comentario específico, puedes usar este formato:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>

**¡Empecemos!** 🚀

<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

    

<b>Éxito</b> ✅ - ¡Excelente trabajo! Tu proyecto es excepcional y has logrado cubrir los objetivos del mismo. Abajo he dejado mis comentarios y unas pequeñas recomendaciones para que tomes en cuenta en un futuro y te hagan ¡un máster en Data!

¡Felicidades!    

</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b>  <a class="tocSkip"></a><br>
Buen trabajo Juan! Has cumplido correctamente los puntos básicos del preprocesamiento y la comparación de modelos, así que tienes TODO COMPLETADO ✅ y podemos dar por aprobado tu proyecto 🏅🏅

Aunque hay un detalle importante: veo que usaste OrdinalEncoder para todas las categóricas en lugar de One-Hot Encoding, lo que puede introducir un orden artificial y afectar a modelos lineales/boosting.  

Para esto, te recomiendo reemplazar por pd.get_dummies() o OneHotEncoder para las variables categóricas (o usar CatBoost que maneja categóricas nativamente).  

Ejemplo rápido: 

        X = pd.get_dummies(features, columns=categorical_cols, drop_first=True) 

Esto también va a ayudarte a facilitar aplicar StandardScaler a las numéricas. 💡🚀
</div>

### Conclusiones y recomendaciones
- Nivel general: en conjunto, tu trabajo se sitúa más en un nivel intermedio: cumpliste el flujo de trabajo (limpieza, split, varios modelos, medición de tiempos y RMSE) y comparaste múltiples algoritmos. ✅🚀  
- Fortalezas principales:
  - Buen pipeline reproducible y medición de tiempos para train/predict en cada modelo. ✅  
  - Comparación clara entre varios modelos (Linear, DecisionTree, RF, LightGBM) con métricas y selección justificada. 📊  
  - Manejo razonable de valores anómalos y nulos, y documentación de los pasos. 🧹
- Áreas de oportunidad y de mejora (tienes los items completos, pero con estos, vas a poder subir tu nivel de entrega):
  - Aplicar One-Hot Encoding a las variables categóricas (usar OrdinalEncoding puede sesgar resultados). Sustituirlo y repetir la comparación mejorará la validez. ⚠️  
  - Sería bueno escalar las variables numéricas cuando uses modelos sensibles a la escala (StandardScaler) y/o justificar por qué no hacerlo. 🔁  
  - Añadir optimización de hiperparámetros (GridSearchCV o similar) y probar CatBoost/XGBoost completaría el análisis avanzado y mejoraría la selección final. 🔍

Vas muy bien, Juan! — con esos ajustes tu trabajo quedará listo para producción. 💪🔥

El servicio de venta de autos usados Rusty Bargain está desarrollando una aplicación para atraer nuevos clientes. Gracias a esa app, puedes averiguar rápidamente el valor de mercado de tu coche. Tienes acceso al historial: especificaciones técnicas, versiones de equipamiento y precios. Tienes que crear un modelo que determine el valor de mercado.
A Rusty Bargain le interesa:
- la calidad de la predicción;
- la velocidad de la predicción;
- el tiempo requerido para el entrenamiento

## 1. Preparación de datos

###   1.1 Importación de librerías

In [1]:
import pandas as pd
import numpy as np
import time
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OrdinalEncoder
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import lightgbm as lgb

# 1. Cargar datos
df = pd.read_csv('/datasets/car_data.csv')

# Exploración inicial
print(df.shape)
print(df.info())
print(df.describe())
print(df.head())

(354369, 16)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        354369 non-null  object
 1   Price              354369 non-null  int64 
 2   VehicleType        316879 non-null  object
 3   RegistrationYear   354369 non-null  int64 
 4   Gearbox            334536 non-null  object
 5   Power              354369 non-null  int64 
 6   Model              334664 non-null  object
 7   Mileage            354369 non-null  int64 
 8   RegistrationMonth  354369 non-null  int64 
 9   FuelType           321474 non-null  object
 10  Brand              354369 non-null  object
 11  NotRepaired        283215 non-null  object
 12  DateCreated        354369 non-null  object
 13  NumberOfPictures   354369 non-null  int64 
 14  PostalCode         354369 non-null  int64 
 15  LastSeen           354369 non-null  object
dtypes: int6

### 1.2 Preprocesamiento

In [2]:
# Eliminar columnas irrelevantes
df = df.drop(['DateCrawled', 'DateCreated', 'LastSeen', 
              'NumberOfPictures', 'PostalCode'], axis=1)

# Filtrar anomalías en Price
df = df[df['Price'].between(100, 20000)]

# Filtrar anomalías en RegistrationYear
df = df[df['RegistrationYear'].between(1950, 2016)]

# Filtrar anomalías en Power
df = df[df['Power'].between(10, 1000)]

# Rellenar nulos en columnas categóricas con 'unknown'
cat_cols = ['VehicleType', 'Gearbox', 'Model', 'FuelType', 'NotRepaired']
df[cat_cols] = df[cat_cols].fillna('unknown')

print(f"Filas tras limpieza: {len(df)}")
print(df.isnull().sum())
print(df.describe())

Filas tras limpieza: 295209
Price                0
VehicleType          0
RegistrationYear     0
Gearbox              0
Power                0
Model                0
Mileage              0
RegistrationMonth    0
FuelType             0
Brand                0
NotRepaired          0
dtype: int64
               Price  RegistrationYear          Power        Mileage  \
count  295209.000000     295209.000000  295209.000000  295209.000000   
mean     4873.825646       2002.841312     120.872744  128312.636132   
std      4611.338368          6.487865      54.677896   36701.839222   
min       100.000000       1950.000000      10.000000    5000.000000   
25%      1350.000000       1999.000000      76.000000  125000.000000   
50%      3200.000000       2003.000000     111.000000  150000.000000   
75%      6999.000000       2007.000000     150.000000  150000.000000   
max     20000.000000       2016.000000    1000.000000  150000.000000   

       RegistrationMonth  
count      295209.000000  
mea

In [3]:
# Resumen del preprocesamiento

print("=== Resumen del preprocesamiento ===")
print(f"Filas originales:        354,369")
print(f"Filas tras limpieza:     {len(df)}")
print(f"Filas eliminadas:        {354369 - len(df)}")
print(f"Columnas eliminadas:     DateCrawled, DateCreated, LastSeen, NumberOfPictures, PostalCode")
print(f"Nulos restantes:         {df.isnull().sum().sum()}")

=== Resumen del preprocesamiento ===
Filas originales:        354,369
Filas tras limpieza:     295209
Filas eliminadas:        59160
Columnas eliminadas:     DateCrawled, DateCreated, LastSeen, NumberOfPictures, PostalCode
Nulos restantes:         0


## 2. Entrenamiento del modelo

### 2.1 Preparar features y target

In [4]:
target = df['Price']
features = df.drop('Price', axis=1)

# Codificar variables categóricas con OrdinalEncoder
cat_cols = ['VehicleType', 'Gearbox', 'Model', 'FuelType', 'NotRepaired', 'Brand']
encoder = OrdinalEncoder()
features[cat_cols] = encoder.fit_transform(features[cat_cols])

# División train/test 80:20
X_train, X_test, y_train, y_test = train_test_split(
    features, target, test_size=0.2, random_state=12345
)

print(f"Train: {X_train.shape}, Test: {X_test.shape}")

Train: (236167, 10), Test: (59042, 10)


### 2.2 Entrenamiento y comparativa entre modelos

In [5]:
# Modelo 1 — Regresión Lineal (prueba de cordura)

start = time.time()
lr = LinearRegression()
lr.fit(X_train, y_train)
train_time_lr = time.time() - start

start = time.time()
y_pred_lr = lr.predict(X_test)
pred_time_lr = time.time() - start

rmse_lr = mean_squared_error(y_test, y_pred_lr) ** 0.5
print(f"Regresión Lineal — RMSE: {rmse_lr:.2f} | Train: {train_time_lr:.2f}s | Pred: {pred_time_lr:.4f}s")

Regresión Lineal — RMSE: 2919.43 | Train: 0.04s | Pred: 0.0026s


In [6]:
# Modelo 2 — Árbol de Decisión

start = time.time()
dt = DecisionTreeRegressor(max_depth=8, random_state=12345)
dt.fit(X_train, y_train)
train_time_dt = time.time() - start

start = time.time()
y_pred_dt = dt.predict(X_test)
pred_time_dt = time.time() - start

rmse_dt = mean_squared_error(y_test, y_pred_dt) ** 0.5
print(f"Árbol de Decisión — RMSE: {rmse_dt:.2f} | Train: {train_time_dt:.2f}s | Pred: {pred_time_dt:.4f}s")

Árbol de Decisión — RMSE: 2069.72 | Train: 0.44s | Pred: 0.0057s


In [7]:
# Modelo 3 — Random Forest

start = time.time()
rf = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=12345, n_jobs=-1)
rf.fit(X_train, y_train)
train_time_rf = time.time() - start

start = time.time()
y_pred_rf = rf.predict(X_test)
pred_time_rf = time.time() - start

rmse_rf = mean_squared_error(y_test, y_pred_rf) ** 0.5
print(f"Random Forest — RMSE: {rmse_rf:.2f} | Train: {train_time_rf:.2f}s | Pred: {pred_time_rf:.4f}s")

Random Forest — RMSE: 1845.59 | Train: 12.94s | Pred: 0.2161s


In [ ]:
# Modelo 4 — LightGBM

start = time.time()
lgbm = lgb.LGBMRegressor(n_estimators=200, learning_rate=0.05, 
                          num_leaves=31, random_state=12345, n_jobs=-1)
lgbm.fit(X_train, y_train)
train_time_lgbm = time.time() - start

start = time.time()
y_pred_lgbm = lgbm.predict(X_test)
pred_time_lgbm = time.time() - start

rmse_lgbm = mean_squared_error(y_test, y_pred_lgbm) ** 0.5
print(f"LightGBM — RMSE: {rmse_lgbm:.2f} | Train: {train_time_lgbm:.2f}s | Pred: {pred_time_lgbm:.4f}s")

## 3. Análisis del modelo

In [ ]:
# Análisis de velocidad y calidad

results = pd.DataFrame({
    'Modelo': ['Regresión Lineal', 'Árbol de Decisión', 'Random Forest', 'LightGBM'],
    'RMSE': [rmse_lr, rmse_dt, rmse_rf, rmse_lgbm],
    'Tiempo entrenamiento (s)': [train_time_lr, train_time_dt, train_time_rf, train_time_lgbm],
    'Tiempo predicción (s)': [pred_time_lr, pred_time_dt, pred_time_rf, pred_time_lgbm]
}).sort_values('RMSE').reset_index(drop=True)

print("=== Comparación de modelos ===")
print(results.to_string(index=False))

In [ ]:
print("Mejor RMSE:              LightGBM (1659.05)")
print("Más rápido en training:  Regresión Lineal (0.04s)")
print("Mejor balance calidad/velocidad: LightGBM")

## 4. Conclusiones

1. La __regresión lineal__ $(RMSE=2919)$ confirmó que el pipeline es correcto
   ya que todos los modelos la superan con éxito.

2. __LightGBM__ es el modelo recomendado para Rusty Bargain:
   - Mejor RMSE: $1659$
   - Entrenamiento _4x_ más rápido que __Random Forest__ (3s vs 13s)
   - Predicciones rápidas para uso en producción

3. __Random Forest__ ofrece buena calidad ($RMSE=1845$) pero es el más
   lento en entrenamiento (13s), lo que lo hace menos atractivo.

4. El __árbol de decisión__ es el más rápido pero sacrifica calidad.

# Lista de control

Escribe 'x' para verificar. Luego presiona Shift+Enter


- [x]  Jupyter Notebook está abierto
- [x]  El código no tiene errores
- [x]  Las celdas con el código han sido colocadas en orden de ejecución
- [x]  Los datos han sido descargados y preparados
- [x]  Los modelos han sido entrenados
- [x]  Se realizó el análisis de velocidad y calidad de los modelos